RecoParticleFlow/PFClusterProducerAlpaka/plugins/alpaka/PFClusterParamsESProducer.cc

```diff
@@ -0,0 +1,253 @@
+#include "FWCore/Framework/interface/EventSetupRecordIntervalFinder.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/Utilities/interface/Exception.h"
+#include "HeterogeneousCore/AlpakaCore/interface/alpaka/ESProducer.h"
+#include "HeterogeneousCore/AlpakaCore/interface/alpaka/ModuleFactory.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/memory.h"
+#include "RecoParticleFlow/PFClusterProducerAlpaka/interface/PFClusterParamsAlpakaESRecord.h"
+#include "RecoParticleFlow/PFClusterProducerAlpaka/interface/alpaka/PFClusterParamsAlpakaESData.h"
+
+#include <array>
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+
+  class PFClusterParamsESProducer : public ESProducer {
+  public:
+    PFClusterParamsESProducer(edm::ParameterSet const& iConfig) : ESProducer(iConfig), paramSet(iConfig) {
+      product = std::make_shared<reco::PFClusterParamsAlpakaESDataHost>(7, cms::alpakatools::host());
```

Comment:
(this question is about me trying to understand better, and any action would be beyond this PR)

All the information to fill this EventSetup data product comes from the configuration, correct? Is the data product foreseen to be consumed by only one EDModule, or possibly many? If only one EDModule, would it be convenient to have this code and the configuration as part of the `PFClusterProducerAlpaka`?

Reply:
I think this use of ES data product came from our earlier CUDA version (which never got integrated into the main CMSSW).

At least when I was working on the CUDA version, I wasn’t sure how to do read various RECO parameters directly in ClusterProducer itself (read various parameters, transfer to devices only once, even when jobs are using multiple CPU threads), so I utilized ES data (and it got propagated to Alpaka version).

If there is a better way to do it, I think it can be considered in the followup PR?

Reply:
I think this is a reasonable approach also when using Alpaka.

Reply:
It is ok for now, but in the longer term the framework team would want to move the "infinite IOV" stuff from (ab)using EventSetup to something more appropriate (e.g. `GlobalCache` for stream modules if the data is needed only for one module, or `ProcessBlock` if needed for many). Facilities to copy the data for all relevant devices should certainly be part of that. I take that in this case only one EDModule consumer is expected.

Reply:
Thank you @makortel . So, am I right that the framework team is working on some facility to utilize `GlobalCache` or `ProcessBlock`, so we should stay tuned?

Reply:
Yes, but don't hold your breath.